ctapipe/calib/camera/calibrator.py

```diff
@@ -152,3 +181,36 @@ def __call__(self, event):
         for telid in tel.keys():
             self._calibrate_dl0(event, telid)
             self._calibrate_dl1(event, telid)
+
+
+def shift_waveforms(waveforms, time_shift_samples):
+    """
+    Shift the waveforms by the mean integer shift to mediate
+    time differences between pixels.
+    The remaining shift (mean + fractional part) is returned
+    to be applied later to the extracted peak time.
+    """
+    mean_shift = time_shift_samples.mean()
+    integer_shift = np.round(time_shift_samples - mean_shift).astype("int16")
+    remaining_shift = time_shift_samples - integer_shift
+    shifted_waveforms = _shift_waveforms_by_integer(waveforms, integer_shift)
+    return shifted_waveforms, remaining_shift
+
+
+@njit
+def _shift_waveforms_by_integer(waveforms, integer_shift):
+    n_pixels, n_samples = waveforms.shape
+    shifted_waveforms = np.zeros_like(waveforms)
+
+    for pixel_idx in range(n_pixels):
+        shift = integer_shift[pixel_idx]
+
+        for new_sample_idx in range(n_samples):
+            # repeat first value if out ouf bounds to the left
+            # repeat last value if out ouf bounds to the right
```

Comment:
If there is a baseline shift (pedestal) it is probably better to repeat.

I would either repeat or set nan, but then we probably would have to deal with nan values properly.

Reply:
I'd rather set "zero" meaning the baseline shift.  But thinking about it, this means that we will also be throwing away (occasionally) samples with perfectly valid values above the noise... this may be very dangerous. A pixel with bad timing but otherwise perfectly healthy would be lost.

Reply:
Either we increase artificially the full range for all pixels (better not), or we drop the "shift approach" and keep the floating point shit value full for use by extractors etc.

Reply:
Without shifting a  bad pixel might kill the good pixels in a NeighborPeak or GlobalPeak.

With shifting, the bad pixel might be killed because it's to far off.

I think I'd go for shifting. 

These shifts should anyway be much smaller than the window width, right? Otherwise the telescope hardware has a real problem that needs to be fixed.

Reply:
What I meant by "for use by extractors" is precisely that a global peak extractor can use the shifts. I agree that the shifts will generally be small, but even so we will always have to kill _some_ 'valid' significant samples, and 'invent' others... Whereas I see no drawback in keeping the shift as a floating-point number that is accessible in DL0 for use by extractors.

Reply:
You expect important significant samples at the edges of the readout windows?

Reply:
Not too often, obviously, but since we are discussing what to do with non-existing samples  when one shifts the pulses, the question comes up of what is the real gain of physically shifting the waveforms rather than just storing the shift value, which can then be used by extractors. The idea, if we shift the waveforms, is that they will be stored that way in DL0. And this means some loss of information (typically small, but if one pixel happens to be badly calibrated in time it may be large). I wonder whether we need to take the risk at all... 
If we agreed to keep the non-integer part of the shift out of the shifting (to avoid resampling), why could we not agree on keeping the _whole_ of the shift as a pixel-wise (&event-wise) value value, and not shift the waveforms physically at all? Also, it would be much cleaner, because if we do the shift we probably want to keep record of what shift was applied... That would be unnecessary if we do not do the shift.

Reply:
Because with this approach, all extractors can benefit from it. Otherwise an extractor like the GlobalPeak or NeighborPeak window sum would have to do the shifting internally and somehow communicate that a part of the shift already happened.

Like this, the implementation is much more straight forward and easier to reason about.

Reply:
> If there is a baseline shift (pedestal) it is probably better to repeat.

Ah, I assumed the pedestal would be subtracted first. Then setting new samples to zero should be acceptable.

Reply:
> Because with this approach, all extractors can benefit from it. Otherwise an extractor like the GlobalPeak or NeighborPeak window sum would have to do the shifting internally 

That does not seem too complicated, right? You add up waveforms now, you add them shifted (by the integer part of the time shift). Am I overlooking something?

> and somehow communicate that a part of the shift already happened.
What do you mean exactly? What we probably would need is to keep track in DL0 of the applied shift to waveforms, if we go for that option.
 
> Like this, the implementation is much more straight forward and easier to reason about.
It is not for free, we will throw away valid samples. I cannot see what is not straightforward about the other approach.

Other opinions? I think we should not throw away values present in the raw data, unless there is clear gain in doing so (like in DVR, which will anyway throw away only insignificant values).

Reply:
But the other approach will throw away the same samples, just in code that is now duplicated in three different places?

Reply:
> The idea, if we shift the waveforms, is that they will be stored that way in DL0. And this means some loss of information (typically small, but if one pixel happens to be badly calibrated in time it may be large). I wonder whether we need to take the risk at all...

Just to be clear, whats being implemented here doesn't effect the waveforms stored at DL0. This is being applied in the DL0 -> DL1 step.

Reply:
And you need to communicate back the amount of shift you already applied, so only the remaining shift is applied to the peak times.

Reply:
Would it be so bad to have an increased waveform size? Based on the maximum shifts specified? (probably can't use guvectorize in that case). This ensures no lost information.

Reply:
> But the other approach will throw away the same samples, just in code that is now duplicated in three different places?

No, why? In particular for 'Local' extractors you have all original samples at hand. For the other extractors, no difference.

Reply:
> > The idea, if we shift the waveforms, is that they will be stored that way in DL0. And this means some loss of information (typically small, but if one pixel happens to be badly calibrated in time it may be large). I wonder whether we need to take the risk at all...
> 
> Just to be clear, whats being implemented here doesn't effect the waveforms stored at DL0. This is being applied in the DL0 -> DL1 step.

Wait... are you sure? I think @maxnoe intends to make this part of the R1 event source. If in your view the R1 & DL0 waveforms are left unshifted, and the shift is just an "internal business" of extractors, or applied just before waveforms are passed to the extractors, then  I have no objection at all (that is actually what I am for)

Reply:
> Would it be so bad to have an increased waveform size? Based on the maximum shifts specified? (probably can't use guvectorize in that case). This ensures no lost information.

Not bad, as long as it is not for writing out waveforms to DL0. Would mean different lengths per run, or even per event (depending on how often the time shifts are recalculated)

Reply:
Hi, @watsonjj I think you are right (about this all being post-DL0)... too many discussions at the same time: there is the parallel issue of moving the low-level LSTcam calibrations to the LST R1 event source, and I thought @maxnoe wanted to include this waveform shifts there. If that is not the case, apologies for the noise!

Reply:
Ah, yes. This is all about the dl1 integration. This happens right before peak extraction and won't affect anything else.

Reply:
ok, thanks! I mixed this with the discussion about the R1 event source.

Reply:
Look at where this code lives. We are in ctapipe (not ctapipe_io_lst) and the only change part is the `_calibrate_dl1` method of the CameraCalibrator, right before the waveforms are passed to `image_extractor`.